doc/read-the-docs-site/tutorials/AuctionValidator.hs

```diff
@@ -0,0 +1,194 @@
+import PlutusLedgerApi.V1.Interval (contains)
+import PlutusLedgerApi.V2 (Datum (..), OutputDatum (..), ScriptContext (..), TxInfo (..), TxOut (..), from, to)
+import PlutusLedgerApi.V2.Contexts (getContinuingOutputs)
+import PlutusTx
+import PlutusTx.Bool
+import PlutusTx.Builtins
+import PlutusTx.Lift
+import PlutusTx.Maybe
+import PlutusTx.Prelude qualified as PlutusTx
+import PlutusTx.Show qualified as PlutusTx
+import Prelude qualified as Haskell
+
+-- BLOCK1
+data AuctionParams = AuctionParams
+  { apSeller  :: PubKeyHash,
+    apAsset   :: Value,
+    apMinBid  :: Integer,
+    apEndTime :: POSIXTime
+  }
+
+PlutusTx.makeLift ''AuctionParams
+
+data Bid = Bid
+  { bBidder :: PubKeyHash,
+    bAmount :: Integer
+  }
+
+PlutusTx.deriveShow ''Bid
+PlutusTx.unstableMakeIsData ''Bid
+
+instance PlutusTx.Eq Bid where
+  {-# INLINEABLE (==) #-}
+  bid == bid' =
+    bBidder bid PlutusTx.== bBidder bid'
+      PlutusTx.&& bAmount bid PlutusTx.== bAmount bid'
+
+newtype AuctionDatum = AuctionDatum { adHighestBid :: Maybe Bid }
+
+PlutusTx.unstableMakeIsData ''AuctionDatum
+
+data AuctionRedeemer = NewBid Bid | Payout
+
+PlutusTx.unstableMakeIsData ''AuctionRedeemer
+-- BLOCK2
+
+
+{-# INLINEABLE auctionTypedValidator #-}
+auctionTypedValidator ::
+  AuctionParams ->
+  AuctionDatum ->
+  AuctionRedeemer ->
+  ScriptContext ->
+  Bool
+auctionTypedValidator params (AuctionDatum highestBid) redeemer ctx@(ScriptContext txInfo _) =
+  PlutusTx.and conditions
+  where
+    conditions :: [Bool]
+    conditions = case redeemer of
+      NewBid bid ->
+        [ -- The new bid must be higher than the highest bid.
+          -- If this is the first bid, it must be at least as high as the minimum bid.
+          sufficientBid bid,
+          -- The bid is not too late.
+          validBidTime,
+          -- The previous highest bid should be refunded.
+          refundsPreviousHighestBid,
+          -- A correct new datum is produced, containing the new highest bid.
+          correctNewDatum bid
+        ]
+      Payout ->
+        [ -- The payout is not too early.
+          validPayoutTime,
+          -- The seller gets the highest bid.
+          sellerGetsHighestBid,
+          -- The highest bidder gets the asset.
+          highestBidderGetsAsset
+        ]
+-- BLOCK3
+    sufficientBid :: Bid -> Bool
+    sufficientBid (Bid _ amt) =
+      maybe
+        (amt PlutusTx.>= apMinBid params)
+        (\(Bid _ amt') -> amt PlutusTx.> amt')
+        highestBid
+-- BLOCK4
+    validBidTime :: Bool
+    validBidTime = to (apEndTime params) `contains` txInfoValidRange txInfo
+-- BLOCK5
+    refundsPreviousHighestBid :: Bool
+    refundsPreviousHighestBid = case highestBid of
+      Nothing -> True
+      Just (Bid bidder amt) ->
+        case PlutusTx.filter
+          (\o -> txOutAddress o PlutusTx.== pubKeyHashAddress bidder)
+          (txInfoOutputs txInfo) of
+          [o] -> txOutValue o PlutusTx.== singleton adaSymbol adaToken amt
```

Comment:
What if I was the previous highest bidder? Then there is likely another output going to my public key with my change from creating the current transaction. Worth a comment about the edge case even if we're not going to handle it. Probably good to draw attention to this sort of thing, approximately any time you make assertions that the list of outputs has some particular shape you need to be quite careful!